In [60]:
# initializing panda
import pandas as pd

In [61]:
# to read csv file
df = pd.read_csv("/content/penguins.csv")

In [62]:
# to print DataFrame
df

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
339  Gentoo     Biscoe             NaN            NaN                NaN   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    MALE  
1         3800.0  FEMALE  
2         3250.0  FEMALE  
3            NaN     NaN  
4         3450.0  FEMALE  
..           ...     ...  
339          NaN     NaN  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE  

[344 rows x 7 columns]

In [63]:
# to previews first 5 rows(by default)
df.head()

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE

In [64]:
# to previews first 10 rows
df.head(10)

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   
5  Adelie  Torgersen            39.3           20.6              190.0   
6  Adelie  Torgersen            38.9           17.8              181.0   
7  Adelie  Torgersen            39.2           19.6              195.0   
8  Adelie  Torgersen            34.1           18.1              193.0   
9  Adelie  Torgersen            42.0           20.2              190.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  
5       3650.0    MALE  
6       3625.0  FEMALE  
7       4675.0    MALE  
8       3475.0     NaN  
9       4250.0     NaN

In [65]:
# to preview last 5 rows(by default)
df.tail()

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
339  Gentoo  Biscoe             NaN            NaN                NaN   
340  Gentoo  Biscoe            46.8           14.3              215.0   
341  Gentoo  Biscoe            50.4           15.7              222.0   
342  Gentoo  Biscoe            45.2           14.8              212.0   
343  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
339          NaN     NaN  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE

In [66]:
# to previews last 2 rows
df.tail(2)

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
342  Gentoo  Biscoe            45.2           14.8              212.0   
343  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
342       5200.0  FEMALE  
343       5400.0    MALE

In [67]:
# random 5 rows
df.sample(5)

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
13      Adelie  Torgersen            38.6           21.2              191.0   
295     Gentoo     Biscoe            48.6           16.0              230.0   
342     Gentoo     Biscoe            45.2           14.8              212.0   
121     Adelie  Torgersen            37.7           19.8              198.0   
166  Chinstrap      Dream            45.9           17.1              190.0   

     body_mass_g     sex  
13        3800.0    MALE  
295       5800.0    MALE  
342       5200.0  FEMALE  
121       3500.0    MALE  
166       3575.0  FEMALE

In [68]:
# random 10 rows with fixed seed
df.sample(5, random_state=1)

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
306  Gentoo  Biscoe            43.4           14.4              218.0   
341  Gentoo  Biscoe            50.4           15.7              222.0   
291  Gentoo  Biscoe            46.4           15.6              221.0   
102  Adelie  Biscoe            37.7           16.0              183.0   
289  Gentoo  Biscoe            50.7           15.0              223.0   

     body_mass_g     sex  
306       4600.0  FEMALE  
341       5750.0    MALE  
291       5000.0    MALE  
102       3075.0  FEMALE  
289       5550.0    MALE

In [69]:
# to know the(no_of_row, no_of_column)
df.shape

(344, 7)

# Data profiling and inspection

In [70]:
df.columns # to see the name of columns

Index(['species', 'island', 'bill_length_mm', 'bill_depth_mm',
       'flipper_length_mm', 'body_mass_g', 'sex'],
      dtype='object')

In [71]:
# to previews data of first 2 rows
df.head(2)

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE

In [72]:
# to describes numericla data
df.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count      342.000000     342.000000         342.000000   342.000000
mean        43.921930      17.151170         200.915205  4201.754386
std          5.459584       1.974793          14.061714   801.954536
min         32.100000      13.100000         172.000000  2700.000000
25%         39.225000      15.600000         190.000000  3550.000000
50%         44.450000      17.300000         197.000000  4050.000000
75%         48.500000      18.700000         213.000000  4750.000000
max         59.600000      21.500000         231.000000  6300.000000

In [73]:
# to check the data type
df.dtypes

species               object
island                object
bill_length_mm       float64
bill_depth_mm        float64
flipper_length_mm    float64
body_mass_g          float64
sex                   object
dtype: object

In [74]:
# to includes  object in describes
df.describe(include='object')

species  island   sex
count      344     344   333
unique       3       3     2
top     Adelie  Biscoe  MALE
freq       152     168   168

# Species Column
* 344 rows
* 3 unique categories
* among 3 unique categories most occuring is Adelie
* with 152 frequency

# Island Column
* 344 rows
* 3 unique categories
* among 3 unique categories most occuring is Biscoe
* with 168 frequency

# Sex Column
* 344 rows
* 2 unique categories
* among 2 unique categories most occuring is Male
* with 168 frequency

In [75]:
# to previews unique data in species column
df['species'].unique()

array(['Adelie', 'Chinstrap', 'Gentoo'], dtype=object)

In [76]:
# to count unique data in species column
df['species'].nunique()

3

In [77]:
# to value count of unique data in species
df['species'].value_counts()

Adelie       152
Gentoo       124
Chinstrap     68
Name: species, dtype: int64

In [78]:
# to previews unique data in island column
df['island'].unique()

array(['Torgersen', 'Biscoe', 'Dream'], dtype=object)

In [93]:
# to count unique data in species column
df['island'].nunique()

3

In [95]:
# to value count of unique data in species
df['island'].value_counts()

Biscoe       168
Dream        124
Torgersen     52
Name: island, dtype: int64

In [81]:
# to previews unique data in sex column
df['sex'].unique()

array(['MALE', 'FEMALE', nan], dtype=object)

In [82]:
# to count unique data in sex column
df['sex'].nunique()

2

In [83]:
# to value count of unique data in species
df['sex'].value_counts()

MALE      168
FEMALE    165
Name: sex, dtype: int64

In [84]:
# to prints information about the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [85]:
# to check if data present null or not
df.isna()

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0      False   False           False          False              False   
1      False   False           False          False              False   
2      False   False           False          False              False   
3      False   False            True           True               True   
4      False   False           False          False              False   
..       ...     ...             ...            ...                ...   
339    False   False            True           True               True   
340    False   False           False          False              False   
341    False   False           False          False              False   
342    False   False           False          False              False   
343    False   False           False          False              False   

     body_mass_g    sex  
0          False  False  
1          False  False  
2          False  False  
3           True   True  
4          False  False  
..           ...    ...  
339         True   True  
340        False  False  
341        False  False  
342        False  False  
343        False  False  

[344 rows x 7 columns]

In [86]:
# to count null presnt in data
df.isna().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64

#` In Species and island no missing data`
# `But, in bill_length, bill_depth, flipper_length and body_mass has same number missing data 2`
#` In sex column there is 11 missing data`

In [87]:
# loc to slice DataFrame
df

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
339  Gentoo     Biscoe             NaN            NaN                NaN   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    MALE  
1         3800.0  FEMALE  
2         3250.0  FEMALE  
3            NaN     NaN  
4         3450.0  FEMALE  
..           ...     ...  
339          NaN     NaN  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE  

[344 rows x 7 columns]

In [88]:
# slice DataFrame according to island with value = "Torgerson"
df['island'] == 'Torgersen'

0       True
1       True
2       True
3       True
4       True
       ...  
339    False
340    False
341    False
342    False
343    False
Name: island, Length: 344, dtype: bool

In [91]:
df.loc[df['island'] == 'Torgersen']

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
5    Adelie  Torgersen            39.3           20.6              190.0   
6    Adelie  Torgersen            38.9           17.8              181.0   
7    Adelie  Torgersen            39.2           19.6              195.0   
8    Adelie  Torgersen            34.1           18.1              193.0   
9    Adelie  Torgersen            42.0           20.2              190.0   
10   Adelie  Torgersen            37.8           17.1              186.0   
11   Adelie  Torgersen            37.8           17.3              180.0   
12   Adelie  Torgersen            41.1           17.6              182.0   
13   Adelie  Torgersen            38.6           21.2              191.0   
14   Adelie  Torgersen            34.6           21.1              198.0   
15   Adelie  Torgersen            36.6           17.8              185.0   
16   Adelie  Torgersen            38.7           19.0              195.0   
17   Adelie  Torgersen            42.5           20.7              197.0   
18   Adelie  Torgersen            34.4           18.4              184.0   
19   Adelie  Torgersen            46.0           21.5              194.0   
68   Adelie  Torgersen            35.9           16.6              190.0   
69   Adelie  Torgersen            41.8           19.4              198.0   
70   Adelie  Torgersen            33.5           19.0              190.0   
71   Adelie  Torgersen            39.7           18.4              190.0   
72   Adelie  Torgersen            39.6           17.2              196.0   
73   Adelie  Torgersen            45.8           18.9              197.0   
74   Adelie  Torgersen            35.5           17.5              190.0   
75   Adelie  Torgersen            42.8           18.5              195.0   
76   Adelie  Torgersen            40.9           16.8              191.0   
77   Adelie  Torgersen            37.2           19.4              184.0   
78   Adelie  Torgersen            36.2           16.1              187.0   
79   Adelie  Torgersen            42.1           19.1              195.0   
80   Adelie  Torgersen            34.6           17.2              189.0   
81   Adelie  Torgersen            42.9           17.6              196.0   
82   Adelie  Torgersen            36.7           18.8              187.0   
83   Adelie  Torgersen            35.1           19.4              193.0   
116  Adelie  Torgersen            38.6           17.0              188.0   
117  Adelie  Torgersen            37.3           20.5              199.0   
118  Adelie  Torgersen            35.7           17.0              189.0   
119  Adelie  Torgersen            41.1           18.6              189.0   
120  Adelie  Torgersen            36.2           17.2              187.0   
121  Adelie  Torgersen            37.7           19.8              198.0   
122  Adelie  Torgersen            40.2           17.0              176.0   
123  Adelie  Torgersen            41.4           18.5              202.0   
124  Adelie  Torgersen            35.2           15.9              186.0   
125  Adelie  Torgersen            40.6           19.0              199.0   
126  Adelie  Torgersen            38.8           17.6              191.0   
127  Adelie  Torgersen            41.5           18.3              195.0   
128  Adelie  Torgersen            39.0           17.1              191.0   
129  Adelie  Torgersen            44.1           18.0              210.0   
130  Adelie  Torgersen            38.5           17.9              190.0   
131  Adelie  Torgersen            43.1           19.

In [90]:
# to slice DataFrame according to island with value = "Biscoe"

df.loc[df['island']== 'Biscoe']

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
20   Adelie  Biscoe            37.8           18.3              174.0   
21   Adelie  Biscoe            37.7           18.7              180.0   
22   Adelie  Biscoe            35.9           19.2              189.0   
23   Adelie  Biscoe            38.2           18.1              185.0   
24   Adelie  Biscoe            38.8           17.2              180.0   
..      ...     ...             ...            ...                ...   
339  Gentoo  Biscoe             NaN            NaN                NaN   
340  Gentoo  Biscoe            46.8           14.3              215.0   
341  Gentoo  Biscoe            50.4           15.7              222.0   
342  Gentoo  Biscoe            45.2           14.8              212.0   
343  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
20        3400.0  FEMALE  
21        3600.0    MALE  
22        3800.0  FEMALE  
23        3950.0    MALE  
24        3800.0    MALE  
..           ...     ...  
339          NaN     NaN  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE  

[168 rows x 7 columns]

In [96]:
# to slice DataFrame according to island with value = "Dream"
df[df['island']=='Dream']

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
30      Adelie  Dream            39.5           16.7              178.0   
31      Adelie  Dream            37.2           18.1              178.0   
32      Adelie  Dream            39.5           17.8              188.0   
33      Adelie  Dream            40.9           18.9              184.0   
34      Adelie  Dream            36.4           17.0              195.0   
..         ...    ...             ...            ...                ...   
215  Chinstrap  Dream            55.8           19.8              207.0   
216  Chinstrap  Dream            43.5           18.1              202.0   
217  Chinstrap  Dream            49.6           18.2              193.0   
218  Chinstrap  Dream            50.8           19.0              210.0   
219  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  
30        3250.0  FEMALE  
31        3900.0    MALE  
32        3300.0  FEMALE  
33        3900.0    MALE  
34        3325.0  FEMALE  
..           ...     ...  
215       4000.0    MALE  
216       3400.0  FEMALE  
217       3775.0    MALE  
218       4100.0    MALE  
219       3775.0  FEMALE  

[124 rows x 7 columns]

In [97]:
# to find mean flipper_length whose species is Adelie
df[df['species']=='Adelie']['flipper_length_mm'].mean()

189.95364238410596

In [98]:
# to find mean bill_length_mm whose species in Chinstrap
df[df['species']=='Chinstrap']['bill_length_mm'].mean()

48.83382352941177

In [121]:
df[(df['sex']== 'MALE') & (df['species']== 'Chinstrap')].bill_length_mm.mean()

51.09411764705882

In [102]:
# to groupby to find bill_length_mm with species and sex
df.groupby(['species','sex']).bill_length_mm.mean()

species    sex   
Adelie     FEMALE    37.257534
           MALE      40.390411
Chinstrap  FEMALE    46.573529
           MALE      51.094118
Gentoo     FEMALE    45.563793
           MALE      49.473770
Name: bill_length_mm, dtype: float64

In [104]:
df.groupby(['species','sex']).mean()

<ipython-input-104-e23b2c751d38>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby(['species','sex']).mean()


bill_length_mm  bill_depth_mm  flipper_length_mm  \
species   sex                                                        
Adelie    FEMALE       37.257534      17.621918         187.794521   
          MALE         40.390411      19.072603         192.410959   
Chinstrap FEMALE       46.573529      17.588235         191.735294   
          MALE         51.094118      19.252941         199.911765   
Gentoo    FEMALE       45.563793      14.237931         212.706897   
          MALE         49.473770      15.718033         221.540984   

                  body_mass_g  
species   sex                  
Adelie    FEMALE  3368.835616  
          MALE    4043.493151  
Chinstrap FEMALE  3527.205882  
          MALE    3938.970588  
Gentoo    FEMALE  4679.741379  
          MALE    5484.836066

# slice rows and column using

` Q. rows from 4, 10 and all column`

In [106]:
df.loc[4:10, :]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
4   Adelie  Torgersen            36.7           19.3              193.0   
5   Adelie  Torgersen            39.3           20.6              190.0   
6   Adelie  Torgersen            38.9           17.8              181.0   
7   Adelie  Torgersen            39.2           19.6              195.0   
8   Adelie  Torgersen            34.1           18.1              193.0   
9   Adelie  Torgersen            42.0           20.2              190.0   
10  Adelie  Torgersen            37.8           17.1              186.0   

    body_mass_g     sex  
4        3450.0  FEMALE  
5        3650.0    MALE  
6        3625.0  FEMALE  
7        4675.0    MALE  
8        3475.0     NaN  
9        4250.0     NaN  
10       3300.0     NaN

In [112]:
# rows from 4, 10 and island from sex
df.loc[4:10, 'island':'sex']


island  bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
4   Torgersen            36.7           19.3              193.0       3450.0   
5   Torgersen            39.3           20.6              190.0       3650.0   
6   Torgersen            38.9           17.8              181.0       3625.0   
7   Torgersen            39.2           19.6              195.0       4675.0   
8   Torgersen            34.1           18.1              193.0       3475.0   
9   Torgersen            42.0           20.2              190.0       4250.0   
10  Torgersen            37.8           17.1              186.0       3300.0   

       sex  
4   FEMALE  
5     MALE  
6   FEMALE  
7     MALE  
8      NaN  
9      NaN  
10     NaN

In [125]:
# to find mean of body_mass_g whose island value = 'Dream', species value= 'Adelie' and sex = 'FEMALE'
df[(df['island']== 'Dream') & (df['species']=='Adelie') & (df['sex']=='FEMALE')].body_mass_g.mean()

3344.4444444444443

# Slicing using iloc

` Q. slice 14 to 19 rows and 1 to 4 columns`

In [134]:
df.iloc[14:20, 1:5]

island  bill_length_mm  bill_depth_mm  flipper_length_mm
14  Torgersen            34.6           21.1              198.0
15  Torgersen            36.6           17.8              185.0
16  Torgersen            38.7           19.0              195.0
17  Torgersen            42.5           20.7              197.0
18  Torgersen            34.4           18.4              184.0
19  Torgersen            46.0           21.5              194.0